File: packages/kits/AriaMcpKit/rust/tests/aria_v2_withheld_tests.rs
use aria_mcp::{dispatcher::Dispatcher, estate_posture::EstatePosture, estate_registry::EstateRegistry,
    jsonrpc::{JSONRPCRequest, JsonValue}};
use locus_kit::{adjectives::AdjectiveSensitivity, drawer_operational::CaptureChannel,
    estate_types::LatticeAnchor, frames::{CaptureFrame, MutationKind}, filter::{RecallFrame, Filter, HydrationLevel},
    tunnel::Tunnel};
use serde_json::{json, Value};
use std::sync::Arc;
use corpus_kit::encoder::{SpanEncoder, EncoderModelSpec, EncoderError, PairScorer, CrossEncoderProfile};

struct WithheldSpanEncoder(EncoderModelSpec);
impl SpanEncoder for WithheldSpanEncoder {
    fn spec(&self) -> &EncoderModelSpec { &self.0 }
    fn encode_query(&self, _: &str) -> Result<Vec<f32>, EncoderError> { Ok(vec![1.0 / 384_f32.sqrt(); 384]) }
    fn encode_spans(&self, spans: &[&str]) -> Result<Vec<Vec<f32>>, EncoderError> {
        Ok(spans.iter().map(|_| vec![1.0 / 384_f32.sqrt(); 384]).collect())
    }
}
struct WithheldPairScorer(CrossEncoderProfile);
impl PairScorer for WithheldPairScorer {
    fn profile(&self) -> &CrossEncoderProfile { &self.0 }
    fn backend(&self) -> &str { "withheld-test" }
    fn score(&self, _: &str, spans: &[&str]) -> Result<Vec<f32>, EncoderError> { Ok(vec![1.0; spans.len()]) }
}
// This fixture measures the explicit five-drawer graph, not the registry's
// automatically appended synthetic containment endpoints.
struct WithheldNoTopology;
impl genius_locus_kit::node_topology::NodeTopologyProvider for WithheldNoTopology {
    fn parent_id(&self, _: &str) -> Option<String> { None }
    fn child_ids(&self, _: &str) -> Vec<String> { vec![] }
    fn tree_edges(&self, _: Option<&[String]>) -> Vec<(String, String)> { vec![] }
}

fn call(dispatcher: &Dispatcher, name: &str, arguments: Value) -> Value {
    let request = JSONRPCRequest::decode(&json!({"jsonrpc":"2.0", "id":1,
        "method":"tools/call", "params":{"name":name,"arguments":arguments}})).unwrap();
    let response = serde_json::to_value(dispatcher.handle(&request)).unwrap();
    assert!(response.get("error").is_none(), "{name}: {response}");
    response["result"].clone()
}
fn check(dispatcher: &Dispatcher, name: &str, args: Value, expected: u64) {
    let off = call(dispatcher, name, args.clone());
    let mut args = args;
    args["report_withheld"] = json!(true);
    let on = call(dispatcher, name, args.clone());
    assert_eq!(on["isError"], false, "{name}: {on}");
    assert!(off["structuredContent"]["meta"].get("withheldBySensitivity").is_none());
    assert_eq!(on["structuredContent"]["meta"]["withheldBySensitivity"], expected, "{name}: {on}");
    assert_eq!(on["structuredContent"]["data"], off["structuredContent"]["data"], "{name}");
    args["report_withheld"] = json!(false);
    let explicit_off = call(dispatcher, name, args);
    assert!(explicit_off["structuredContent"]["meta"].get("withheldBySensitivity").is_none());
    assert_eq!(explicit_off["structuredContent"]["data"], off["structuredContent"]["data"]);
}

#[test]
fn report_withheld_shipping_v2() {
    const NOW: i64 = 1_780_000_000_000;
    let mut registry = EstateRegistry::new_inmemory();
    let handle = registry.default.handle.clone();
    let shared = registry.coord.clone();
    shared.lock().unwrap().register_node_topology(&handle, Arc::new(WithheldNoTopology));
    let mut ids = Vec::new();
    {
        let coordinator = shared.lock().unwrap();
        for index in 0..5 {
            let drawer = coordinator.capture(&handle, CaptureFrame::new(
                format!("withheld probe document {index} has distinct useful evidence"),
                CaptureChannel::Typed, "r", LatticeAnchor::udc("004"), "withheld-v2", "test-v1"), NOW + index).unwrap();
            ids.push(drawer.id);
        }
        let estate = coordinator.estate_for(&handle).unwrap();
        for (index, leaf) in ids.iter().enumerate().skip(1) {
            let mut tunnel = Tunnel::new(format!("edge-{index}"), "withheld-graph".into(), "r".into(),
                "withheld-graph".into(), "r".into(), "relates".into(), "withheld-v2".into(), NOW);
            tunnel.source_drawer_id = Some(ids[0].clone());
            tunnel.target_drawer_id = Some(leaf.clone());
            estate.add_tunnel(&tunnel).unwrap();
        }
        estate.mutate(&ids[0], MutationKind::CorrectSensitivity(AdjectiveSensitivity::Restricted), None).unwrap();
        estate.mutate(&ids[1], MutationKind::CorrectSensitivity(AdjectiveSensitivity::Restricted), None).unwrap();
    }
    let peer_id = registry.register_inmemory("withheld-peer");
    let peer = registry.resolve(&std::collections::BTreeMap::from([
        ("estateID".to_owned(), JsonValue::String(peer_id.to_string()))]), "estateID").unwrap().handle.clone();
    {
        use genius_locus_kit::{GrantOptions, GrantScope, CustodyMode, GrantLifetime, ReSharePermission};
        let mut coordinator = shared.lock().unwrap();
        let issued = coordinator.issue_grant(&peer, GrantOptions {
            grantee_estate_id: uuid::Uuid::from_bytes(handle.estate_uuid),
            scope: GrantScope::WholeEstate, custody_mode: CustodyMode::HandedOver,
            lifetime: GrantLifetime::Permanent, content_level: 48, re_share_permission: ReSharePermission::None,
        }, &[0xD1; 32], NOW as f64 / 1000.0).unwrap();
        coordinator.grant_store_mut(&peer).unwrap().set_budget(issued.grant.id, 1.0).unwrap();
        let mut capture = CaptureFrame::new("peer restricted probe", CaptureChannel::Typed,
            "r", LatticeAnchor::udc("004"), "withheld-v2", "test-v1");
        capture.sensitivity = AdjectiveSensitivity::Restricted;
        coordinator.capture(&peer, capture, NOW).unwrap();
    }
    let dispatcher = Dispatcher::new(registry, "aria-mcp-test", "test", "test-serial", None)
        .with_posture(EstatePosture::Live);
    let ranked = call(&dispatcher, "moot_lens_keystones", json!({"wing":"withheld-graph","topK":"1"}));
    assert_eq!(ranked["structuredContent"]["data"]["keystones"][0]["id"], ids[0], "fixture must rank its restricted hub first");
    {
        let coordinator = shared.lock().unwrap();
        let frame = RecallFrame::new(vec![]);
        assert_eq!(coordinator.hydrate_with_sensitivity_count(&handle, &ids[..1], &frame).unwrap().withheld_by_sensitivity, 1);
    }
    let cases = [
        ("moot_recall_precise", json!({"query":"withheld probe"}), 2),
        ("moot_recall_shaped", json!({"query":"withheld probe"}), 2),
        ("moot_recall_connected", json!({"query":"withheld probe"}), 2),
        ("moot_recall_distilled", json!({"query":"withheld probe"}), 2),
        ("moot_recall_vague", json!({"query":"withheld probe"}), 0),
        ("moot_lens_partial_cue", json!({"anchor_memory_id":ids[2]}), 2),
        ("moot_lens_trust_synthesis", json!({}), 2),
        ("moot_lens_keystones", json!({"wing":"withheld-graph","topK":"1"}), 1),
        ("moot_lens_keystones", json!({"wing":"withheld-graph","topK":"5"}), 2),
        ("moot_federated_recall", json!({"filter":"unconfirmed","hydration_level":"full"}), 1),
    ];
    for (name, args, expected) in cases { check(&dispatcher, name, args, expected); }
    {
        let coordinator = shared.lock().unwrap();
        coordinator.withdraw(&handle, &ids[0], Some("obsolete"), NOW + 10).unwrap();
        let mut frame = RecallFrame::new(vec![]); frame.hydration_level = HydrationLevel::Full;
        let result = coordinator.hydrate_with_sensitivity_count(&handle, &ids, &frame).unwrap();
        assert_eq!(result.withheld_by_sensitivity, 1);
        assert!(!result.drawers.iter().any(|d| d.id == ids[0] || d.id == ids[1]));
        frame.filter_chain = vec![Filter::SensitivityAtMost(AdjectiveSensitivity::Secret)];
        assert_eq!(coordinator.hydrate_with_sensitivity_count(&handle, &ids, &frame).unwrap().withheld_by_sensitivity, 0);
    }
    check(&dispatcher, "moot_lens_keystones", json!({"wing":"withheld-graph","topK":"1"}), 0);
    let transcript = call(&dispatcher, "moot_memory_recall_transcript", json!({"query":"withheld probe","report_withheld":true}));
    assert_eq!(transcript["isError"], true);
    assert!(transcript["structuredContent"]["meta"].get("withheldBySensitivity").is_none());
    // Use real Synapse serving receipts; only inference is deterministic.
    {
        use genius_locus_kit::coordinator::EstateCoordinator;
        use genius_locus_kit::span_rerank::{encoder_spec_from_row, SpanEncoderQuerySeam, SynapseSpanVectorReader};
        use persistence_kit::inmemory::InMemoryStorage;
        use synapsekit::vector_store::{VectorStore, SpanVectorInput};
        let mut coordinator = shared.lock().unwrap();
        let content = "User: withheld transcript probe alpha beta gamma delta\nAssistant: retained evidence epsilon zeta eta theta";
        let drawer = coordinator.capture(&handle, CaptureFrame::new(content, CaptureChannel::Typed,
            "r", LatticeAnchor::udc("004"), "withheld-v2", "test-v1"), NOW + 20).unwrap();
        let corpus = Arc::new(corpus_kit::CorpusContentEngine::standalone_on(
            Arc::new(InMemoryStorage::with_estate(uuid::Uuid::new_v4())),
            vec![corpus_kit::EmbeddingModelConfig::Deterministic]).unwrap());
        corpus.ingest(content, &drawer.id, NOW).unwrap();
        coordinator.register_corpus(&handle, corpus);
        let vectors = Arc::new(VectorStore::open(Arc::new(InMemoryStorage::with_estate(uuid::Uuid::new_v4()))).unwrap());
        let record = EstateCoordinator::default_encoder_model_row(true);
        let encoder: Arc<dyn SpanEncoder> = Arc::new(WithheldSpanEncoder(encoder_spec_from_row(&record)));
        coordinator.seed_default_encoder_model_if_absent(&handle).unwrap();
        coordinator.register_span_encoder(&handle, encoder.clone());
        coordinator.register_span_rerank(&handle, Arc::new(SpanEncoderQuerySeam(encoder)),
            Arc::new(SynapseSpanVectorReader(vectors.clone())), 30);
        coordinator.register_pair_scorer(&handle, Arc::new(WithheldPairScorer(CrossEncoderProfile::minilm_l6())));
        vectors.write_span_vectors(&drawer.id, &record.model_id, &record.model_version,
            &(0..3).map(|index| SpanVectorInput { index, int8: vec![1; 384], scale: 0.01,
                start_word: index as usize * 4, end_word: (index as usize + 1) * 4,
                content_version: genius_locus_kit::span_content_version::span_content_version(content),
            }).collect::<Vec<_>>(), NOW).unwrap();
    }
    // Like Swift, transcript supplies an explicit caller ceiling. The default-
    // ceiling-only count is zero even with the restricted candidate present.
    check(&dispatcher, "moot_memory_recall_transcript", json!({"query":"withheld transcript probe"}), 0);
    let help = call(&dispatcher, "moot_help", json!({"report_withheld":true}));
    assert!(help["structuredContent"]["meta"].get("withheldBySensitivity").is_none());
    let tools = aria_mcp::v2::catalog::selected_tools();
    assert!(tools.as_array().unwrap().iter().all(|op| op["inputSchema"]["properties"].get("report_withheld").is_none()));
    let artifact: Value = serde_json::from_str(include_str!("../../Registry/aria-v2-selected-release.json")).unwrap();
    assert_eq!(artifact["catalogIdentity"], aria_mcp::v2::catalog::selected_capability_digest());
    let fixture: Value = serde_json::from_str(include_str!("../../Tests/Conformance/global_modifiers_help_fixture.json")).unwrap();
    assert_eq!(fixture["expected"], aria_mcp::v2::help::GLOBAL_MODIFIERS_HELP_TEXT);

    // An otherwise-matching restricted candidate outside the requested wing
    // contributes zero; adding one inside that wing increments the count.
    let wing_arguments = json!({"query":"connected wing probe", "wing":"Withheld A"});
    {
        let coordinator = shared.lock().unwrap();
        for (wing, sensitivity) in [("Withheld A", AdjectiveSensitivity::Normal), ("Withheld B", AdjectiveSensitivity::Restricted)] {
            let mut capture = CaptureFrame::new("connected wing probe evidence", CaptureChannel::Typed,
                "r", LatticeAnchor::udc("004"), "withheld-v2", "test-v1");
            capture.wing = Some(wing.to_owned());
            capture.sensitivity = sensitivity;
            coordinator.capture(&handle, capture, NOW + 30).unwrap();
        }
    }
    check(&dispatcher, "moot_recall_connected", wing_arguments.clone(), 0);
    {
        let coordinator = shared.lock().unwrap();
        let mut capture = CaptureFrame::new("connected wing probe evidence", CaptureChannel::Typed,
            "r", LatticeAnchor::udc("004"), "withheld-v2", "test-v1");
        capture.wing = Some("Withheld A".to_owned());
        capture.sensitivity = AdjectiveSensitivity::Restricted;
        coordinator.capture(&handle, capture, NOW + 31).unwrap();
    }
    check(&dispatcher, "moot_recall_connected", wing_arguments, 1);
}

File: packages/kits/AriaMcpKit/Tests/AriaMCPTests/AriaV2WithheldTests.swift
import AriaMCPWire
import CorpusKit
import EngramLib
import Foundation
@testable import GeniusLocusKit
import LocusKit
import PersistenceKit
import PersistenceKitInMemory
import SynapseKit
import Testing
@testable import AriaMCP

/// Constant-value float provider: returns the same 384-d vector for every
/// non-empty input. Used to populate the dense lane in tests that do not
/// exercise semantic similarity.
private struct ConstantFloatProvider: EmbeddingProvider, @unchecked Sendable {
    let modelID: String
    let modelVersion = "1.0.0"
    let value: Float

    func embed(_ text: String) async throws -> Engram { .zero }

    func embedFloat(_ text: String) async throws -> [Float] {
        guard !text.isEmpty else { return [] }
        return Array(repeating: value, count: 384)
    }
}

@Suite("Sensitivity-only counts at the shipping v2 door", .serialized)
struct AriaV2WithheldTests {
    @Test func reportWithheldShippingV2() async throws {
        let kit = GeniusLocusKit()
        let owner = OwnerCredentials(ownerIdentifier: "withheld-v2")
        let storage = InMemoryStorage(configuration: EstateConfiguration(estateID: UUID(), backend: .inMemory))
        _ = try await LocusKit.Estate.create(storage: storage, owner: owner)
        let handle = try await kit.open(storage: storage, owner: owner,
            identityKeyStore: InMemoryEstateIdentityKeyStore(), federate: true)
        var ids: [String] = []
        for index in 0..<5 {
            ids.append(try await kit.capture(handle, CaptureFrame(
                content: "withheld probe document \(index) has distinct useful evidence",
                channel: .typed, room: "r", latticeAnchor: .udc("004"),
                addedBy: "withheld-v2", embeddingModelID: "test-v1", subject: "probe \(index)")).id)
        }
        for leaf in ids.dropFirst() {
            _ = try await kit.captureTunnel(handle, TunnelCaptureFrame(
                sourceWing: "study", sourceRoom: "r", targetWing: "study", targetRoom: "r",
                label: "relates", addedBy: "withheld-v2", sourceDrawerId: ids[0], targetDrawerId: leaf,
                kind: .references))
        }
        // Stale normal tunnels can name drawers subsequently reclassified.
        try await kit.mutate(handle, MutateFrame(rowID: ids[0], kind: .correctSensitivity(.restricted)))
        try await kit.mutate(handle, MutateFrame(rowID: ids[1], kind: .correctSensitivity(.restricted)))
        let dispatcher = ToolDispatcher(kit: kit, handle: handle)

        func count(_ result: JSONValue) -> Int64? {
            result.objectValue?["structuredContent"]?.objectValue?["meta"]?.objectValue?["withheldBySensitivity"]?.integerValue
        }
        func data(_ result: JSONValue) -> JSONValue? {
            result.objectValue?["structuredContent"]?.objectValue?["data"]
        }
        func check(_ dispatcher: ToolDispatcher, _ name: String, _ args: [String: JSONValue], _ expected: Int64) async throws {
            let off = try await dispatcher.dispatch(name: name, arguments: .object(args))
            var onArgs = args; onArgs["report_withheld"] = .bool(true)
            let on = try await dispatcher.dispatch(name: name, arguments: .object(onArgs))
            #expect(on.objectValue?["isError"] == .bool(false), "\(name): \(on)")
            #expect(count(off) == nil)
            #expect(count(on) == expected, "\(name): \(on)")
            #expect(data(on) == data(off), "modifier changed rows for \(name)")
            onArgs["report_withheld"] = .bool(false)
            let explicitOff = try await dispatcher.dispatch(name: name, arguments: .object(onArgs))
            #expect(count(explicitOff) == nil)
            #expect(data(explicitOff) == data(off))
        }
        let cases: [(String, [String: JSONValue], Int64)] = [
            ("moot_recall_precise", ["query": .string("withheld probe")], 2),
            ("moot_recall_shaped", ["query": .string("withheld probe")], 2),
            ("moot_recall_connected", ["query": .string("withheld probe")], 2),
            ("moot_recall_distilled", ["query": .string("withheld probe")], 2),
            ("moot_recall_vague", ["query": .string("withheld probe")], 0),
            ("moot_lens_partial_cue", ["anchor_memory_id": .string(ids[2].lowercased())], 2),
            ("moot_lens_trust_synthesis", [:], 2),
            ("moot_lens_keystones", ["wing": .string("study"), "topK": .string("1")], 1),
            ("moot_lens_keystones", ["wing": .string("study"), "topK": .string("5")], 2),
        ]
        for (name, args, expected) in cases { try await check(dispatcher, name, args, expected) }
        try await kit.withdraw(handle, WithdrawFrame(rowID: ids[0], reason: "obsolete"))
        try await check(dispatcher, "moot_lens_keystones", ["wing": .string("study"), "topK": .string("1")], 0)
        let hydrated = try await kit.hydrateWithSensitivityCount(handle, ids: ids,
            frame: RecallFrame(filterChain: []), hydrationLevel: .full)
        #expect(hydrated.withheldBySensitivity == 1)
        #expect(!hydrated.drawers.contains { $0.id == ids[0] || $0.id == ids[1] })
        let explicit = try await kit.hydrateWithSensitivityCount(handle, ids: ids,
            frame: RecallFrame(filterChain: [.sensitivityAtMost(.secret)]), hydrationLevel: .full)
        #expect(explicit.withheldBySensitivity == 0)

        let peerStorage = InMemoryStorage(configuration: EstateConfiguration(estateID: UUID(), backend: .inMemory))
        _ = try await LocusKit.Estate.create(storage: peerStorage, owner: owner)
        let peer = try await kit.open(storage: peerStorage, owner: owner,
            identityKeyStore: InMemoryEstateIdentityKeyStore(), federate: true)
        _ = try await kit.issueGrant(peer, GrantOptions(granteeEstateID: handle.estateUUID,
            scope: .wholeEstate, custodyMode: .mediated, lifetime: .permanent, contentLevel: 48))
        _ = try await kit.capture(peer, CaptureFrame(content: "peer restricted probe", channel: .typed,
            room: "r", latticeAnchor: .udc("004"), addedBy: "withheld-v2", embeddingModelID: "test-v1",
            sensitivity: .restricted))
        try await check(dispatcher.registering(peer), "moot_federated_recall",
            ["filter": .string("unconfirmed"), "hydration_level": .string("full")], 1)

        // A missing required transcript stage remains a refusal even when the
        // modifier is supplied. Its count must not become an existence oracle.
        let transcript = try await dispatcher.dispatch(name: "moot_memory_recall_transcript",
            arguments: .object(["query": .string("withheld probe"), "report_withheld": .bool(true)]))
        #expect(transcript.objectValue?["isError"] == .bool(true))
        #expect(count(transcript) == nil)
        // Exercise successful strict recall through the same production door,
        // with a real serving snapshot and deterministic inference seams.
        let content = "User: withheld transcript probe alpha beta gamma delta\nAssistant: retained evidence epsilon zeta eta theta"
        let drawer = try await kit.capture(handle, CaptureFrame(content: content, channel: .typed,
            room: "r", latticeAnchor: .udc("004"), addedBy: "withheld-v2", embeddingModelID: "test-v1"))
        let corpusStorage = InMemoryStorage(configuration: EstateConfiguration(estateID: UUID(), backend: .inMemory))
        let corpus = try await CorpusContentEngine(standaloneOn: corpusStorage,
            models: [.lsa(provider: ConstantFloatProvider(modelID: "test-miniLM-v1", value: 0.05))])
        try await corpus.ingest(content, contentID: drawer.id, now: Date())
        await kit.registerCorpus(corpus, for: handle)
        let vectorStorage = InMemoryStorage(configuration: EstateConfiguration(estateID: UUID(), backend: .inMemory))
        try await vectorStorage.open(schema: VectorStore.schemaDeclaration)
        let vectors = VectorStore(storage: vectorStorage)
        let record = GeniusLocusKit.defaultEncoderModelRow(isActive: true)
        let encoder = WithheldSpanEncoder(spec: EncoderModelSpec(row: record))
        try await kit.seedDefaultEncoderModelIfAbsent(for: handle)
        await kit.registerSpanEncoder(encoder, for: handle)
        await kit.registerSpanRerank(SpanEncoderQuerySeam(encoder: encoder),
            spanVectors: SynapseSpanVectorReader(store: vectors), for: handle)
        await kit.registerPairScorer(WithheldPairScorer(), for: handle)
        try await vectors.writeSpanVectors(itemID: drawer.id, modelID: record.modelID,
            modelVersion: record.modelVersion, spans: (0..<3).map { index in
                SpanVectorInput(index: UInt32(index), int8: Array(repeating: 1, count: 384), scale: 0.01,
                    startWord: index * 4, endWord: (index + 1) * 4,
                    contentVersion: SpanContentVersion.fnv1a64(content))
            }, filedAt: Date())
        // This recipe already has an explicit sensitivity predicate: Unit 1
        // therefore defines its default-ceiling-only count as zero.
        try await check(dispatcher, "moot_memory_recall_transcript", ["query": .string("withheld transcript probe")], 0)
        let unrelated = try await dispatcher.dispatch(name: "moot_help", arguments: .object(["report_withheld": .bool(true)]))
        #expect(count(unrelated) == nil)
        #expect(AriaV2SelectedCatalog.registry(environment: [:]).operations.allSatisfy {
            $0.inputSchema.objectValue?["properties"]?.objectValue?["report_withheld"] == nil
        })
        try AriaV2CapabilityDigestTests().selectedCatalogDigestSharedVector()
        try SessionProtocolTests().globalModifiersHelpByteIdentity()

        // Wing A excludes otherwise-matching restricted candidates in wing B.
        // Adding the same restricted content inside A then increments only A.
        let wingArguments: [String: JSONValue] = [
            "query": .string("connected wing probe"), "wing": .string("Withheld A")]
        for (wing, sensitivity) in [("Withheld A", AdjectiveSensitivity.normal), ("Withheld B", .restricted)] {
            _ = try await kit.capture(handle, CaptureFrame(content: "connected wing probe evidence",
                channel: .typed, room: "r", latticeAnchor: .udc("004"), addedBy: "withheld-v2",
                embeddingModelID: "test-v1", sensitivity: sensitivity, wing: wing))
        }
        try await check(dispatcher, "moot_recall_connected", wingArguments, 0)
        _ = try await kit.capture(handle, CaptureFrame(content: "connected wing probe evidence",
            channel: .typed, room: "r", latticeAnchor: .udc("004"), addedBy: "withheld-v2",
            embeddingModelID: "test-v1", sensitivity: .restricted, wing: "Withheld A"))
        try await check(dispatcher, "moot_recall_connected", wingArguments, 1)
    }
}

private struct WithheldSpanEncoder: SpanEncoder {
    let spec: EncoderModelSpec
    func encodeQuery(_ text: String) async throws -> [Float] {
        Array(repeating: Float(1) / Float(384).squareRoot(), count: 384)
    }
    func encodeSpans(_ spans: [String]) async throws -> [[Float]] {
        spans.map { _ in Array(repeating: Float(1) / Float(384).squareRoot(), count: 384) }
    }
}

private struct WithheldPairScorer: PairScorer {
    let profile: CrossEncoderProfile = .minilmL6
    var backend: String { "withheld-test" }
    func score(query: String, spans: [String]) async throws -> [Float] {
        Array(repeating: 1, count: spans.count)
    }
}
